### ISSUE TITLE
Memory leak on Apple Silicon (MPS) - missing torch.no_grad()

### ISSUE THREAD

**Issue by chaye7417** [ElectricAlexis/NotaGen#45]
## Problem

Running inference on Apple Silicon (M1/M2/M3/M4) with MPS backend causes memory usage to grow to 90GB+, leading to system freeze.

**Environment:**
- macOS 14+
- Apple M4 Max, 64GB RAM
- PyTorch with MPS backend

## Root Cause

The `generate()` methods in `utils.py` are missing `@torch.no_grad()` decorator, causing gradient graphs to accumulate during inference.

## Fix

Add `@torch.no_grad()` to both generate methods in `utils.py`:

```python
# Line ~308 (CharLevelDecoder.generate)
@torch.no_grad()
def generate(self, encoded_patch, tokens):
    ...

# Line ~383 (NotaGenLMHeadModel.generate)  
@torch.no_grad()
def generate(self, patches, top_k=0, top_p=1, temperature=1.0):
    ...
```

Also recommend:
1. Load checkpoint to CPU first, then move to device
2. Delete checkpoint after loading to free memory
3. Call `torch.mps.empty_cache()` periodically

## Result

After fix:
- Memory usage: stable at 2-3GB
- No more system freeze

I've created a macOS-optimized fork with these fixes: https://github.com/chaye7417/NotaGen-macOS

**Comment by chaye7417** [comment]
抱歉，我重新用中文提交一个 issue
